### PR TITLE
Multi trade workorders

### PR DESCRIPTION
**Pull request by LBHCallumM**
## Link to JIRA ticket

[Story](https://www.pivotaltracker.com/story/show/181056484)

## Describe this PR

Allow a user to vary a workOrder with SOR codes from multiple trades. 

The list below contains which trades should be related to which.
https://docs.google.com/spreadsheets/d/1ebsVMmyrjwqfRGFGI0JTOnT7uD5Qf40-UP53sZm1I3E/edit#gid=1367481187

===

- Added the table 'RelatedTrade' to link between a primary trade, and its related trades
- Additional trades should only be shown when varying a workOrder. I have added the property `showAdditionalTrades` to the request object.

Request object:
```
{
  propertyReference: "",
  contractorReference : "",
  isRaisable : false,
  showAdditionalTrades : false
}
```

Affected endpoints:
`/api/schedule-of-rates/codes`

The changes won't need a feature toggle, because the property `showAdditionalTrades` in the request defaults to false.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [ ] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ ] Checked all code for possible refactoring
- [ ] Code pipeline builds correctly

### CODE REVIEW

**Review by teodor-hristoforov-hackney — APPROVED** [review]

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 1]
```diff
@@ -148,12 +158,21 @@ public async Task ValidatesPropertyRef()
             var invalidContracts = await SeedContracts("not" + expectedProperty, DateTime.UtcNow.AddDays(-7), DateTime.UtcNow.AddDays(7), contractorReference);
             var validContracts = await SeedContracts(expectedProperty, DateTime.UtcNow.AddDays(-7), DateTime.UtcNow.AddDays(7), contractorReference);
 
+            var request = new ListScheduleOfRatesRequest
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:161]: DRY - extract to a helper method (even within the same class)

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 2]
```diff
@@ -168,12 +187,20 @@ public async Task ValidatesTradeCode()
             var notExpectedTrade = await SeedTrade(Guid.NewGuid().ToString());
             var validContracts = await SeedContracts(expectedProperty, DateTime.UtcNow.AddDays(-7), DateTime.UtcNow.AddDays(7), contractorReference);
 
+            var request = new ListScheduleOfRatesRequest
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:190]: As above

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 3]
```diff
@@ -187,12 +214,20 @@ public async Task OnlyReturnsWhereIsRaisableIsTrue()
             var expectedTrade = await SeedTrade(Guid.NewGuid().ToString());
             var validContracts = await SeedContracts(expectedProperty, DateTime.UtcNow.AddDays(-7), DateTime.UtcNow.AddDays(7), contractorReference);
 
+            var request = new ListScheduleOfRatesRequest
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:217]: As above

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 4]
```diff
@@ -206,12 +241,135 @@ public async Task OnlyReturnsWhereIsRaisableIsFalse()
...
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:244]: As above

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 5]
```diff
@@ -128,12 +130,20 @@ public async Task ValidatesContractDates()
             var invalidContracts = await SeedContracts(expectedProperty, DateTime.UtcNow.AddDays(-14), DateTime.UtcNow.AddDays(-7), contractorReference);
             var validContracts = await SeedContracts(expectedProperty, DateTime.UtcNow.AddDays(-7), DateTime.UtcNow.AddDays(7), contractorReference);
 
+            var request = new ListScheduleOfRatesRequest
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:133]: As below

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 6]
```diff
@@ -206,12 +241,135 @@ public async Task OnlyReturnsWhereIsRaisableIsFalse()
...
+            var secondaryTradeSorCodes = await SeedSorCodes(null, null, secondaryTrade, validContracts.First(), true, true);
+
+            await InMemoryDb.Instance.SaveChangesAsync();
+
+            var request = new ListScheduleOfRatesRequest
+            {
+                TradeCode = primaryTrade.Code,
+                ContractorReference = contractorReference,
+                PropertyReference = expectedProperty,
+                ShowAdditionalTrades = true,
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:282]: shouldn't this be false?

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 7]
```diff
@@ -206,12 +241,135 @@ public async Task OnlyReturnsWhereIsRaisableIsFalse()
...
+            };
+
+            // Act
+            var response = await _classUnderTest.GetSorCodes(request);
+
+            // Assert
+            response.Should().HaveCount(primaryTradeSorCodes.Count);
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:289]: Add a check that there are no codes from other trades

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 8]
```diff
@@ -206,12 +241,135 @@ public async Task OnlyReturnsWhereIsRaisableIsFalse()
...
+                ContractorReference = contractorReference,
+                PropertyReference = expectedProperty,
+                ShowAdditionalTrades = false,
+            };
+
+            // Act
+            var response = await _classUnderTest.GetSorCodes(request);
+
+            // Assert
+            response.Should().HaveCount(primaryTradeSorCodes.Count);
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:361]: If this is the test for not showing related trades - Add a check that there are no codes from other trades

File: RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs [hunk 9]
```diff
@@ -225,12 +383,20 @@ public async Task ReturnsAllIfIsRaisableIsNull()
             var expectedTrade = await SeedTrade(Guid.NewGuid().ToString());
             var validContracts = await SeedContracts(expectedProperty, DateTime.UtcNow.AddDays(-7), DateTime.UtcNow.AddDays(7), contractorReference);
 
+            var request = new ListScheduleOfRatesRequest
```
Comment (Asura5) [inline RepairsApi.Tests/V2/Gateways/ScheduleOfRatesGatewayTests.cs:386]: Could be in the helper described earlier

**Review by Asura5 — CHANGES_REQUESTED** [review]
See comments

**Comment by LBHCallumM** [comment]
> See comments

Thanks. I have created a helper method to build the request object, and refactored any code that is reused.

**Review by Asura5 — APPROVED** [review]

**Pull request merged**